src/osp/CoordinateSpaces/CartesianSimple.cpp

```diff
@@ -0,0 +1,82 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "CartesianSimple.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+using namespace osp::universe;
+
+void CoordspaceCartesianSimple::update_exchange(
+        Universe& rUni, CoordinateSpace& rSpace,
+        CoordspaceCartesianSimple& rData)
+{
+    rData.reserve(rData.size() + rSpace.m_toAdd.size());
+
+    auto view = rUni.get_reg().view<UCompCoordspaceIndex>();
+
+    for (auto const &[sat, pos, vel] : rSpace.m_toAdd)
+    {
+        auto &rCoordIndex = view.get<UCompCoordspaceIndex>(sat);
+        rCoordIndex.m_myIndex = rData.size();
+
+        rData.m_satellites.push_back(sat);
+        rData.m_positions.push_back(pos);
+        rData.m_velocities.push_back(vel);
+    }
+
+    rSpace.m_toAdd.clear();
+
+    // TODO: handle removal of satellites
+}
+
+void CoordspaceCartesianSimple::update_views(
+        CoordinateSpace& rSpace, CoordspaceCartesianSimple& rData)
+{
+
+    size_t const ccompCount = std::max({
+                ccomp_id<CCompSat>(),
+                ccomp_id<CCompX>(), ccomp_id<CCompY>(), ccomp_id<CCompZ>()});
```

Comment:
Could you elaborate on the logic here? CCompSat seems like a different unit than the others.

Reply:
This is calculating the size needed for the m_components vector, so each ccomp_id is a valid index. I should probably add a comment for this.

Reply:
Found a better solution using a template function; will be using this instead.
```cpp
template<typename ... CCOMP_T>
constexpr size_t ccomp_min_size() noexcept
{
    return size_t(std::max({ccomp_id<CCOMP_T>() ...})) + 1;
}
```

Reply:
that'd work

you can also make this an inline constexpr template variable, no need for the function syntax.

6-of-one, half-dozen-of-other.